packages/functional-tests/src/app.ts

```diff
@@ -3,203 +3,252 @@
  * Licensed under the MIT License.
  */
 
-import * as MRESDK from '@microsoft/mixed-reality-extension-sdk';
+import * as MRE from '@microsoft/mixed-reality-extension-sdk';
 import * as MRERPC from '@microsoft/mixed-reality-extension-sdk/built/rpc';
-import AltspaceVRLibraryTest from './tests/altspacevr-library-test';
-import AltspaceVRVideoTest from './tests/altspacevr-video-test';
-import AssetEarlyAssignmentTest from './tests/asset-early-assignment-test';
-import AssetMutabilityTest from './tests/asset-mutability-test';
-import AssetPreloadTest from './tests/asset-preload-test';
-import ClockSyncTest from './tests/clock-sync-test';
-import GltfAnimationTest from './tests/gltf-animation-test';
-import GltfConcurrencyTest from './tests/gltf-concurrency-test';
-import GltfGenTest from './tests/gltf-gen-test';
-import InputTest from './tests/input-test';
-import InterpolationTest from './tests/interpolation-test';
-import LightTest from './tests/light-test';
-import LookAtTest from './tests/look-at-test';
-import PrimitivesTest from './tests/primitives-test';
-import ReparentTest from './tests/reparent-test';
-import RigidBodyTest from './tests/rigid-body-test';
-import Test from './tests/test';
-import TextTest from './tests/text-test';
-import UserTest from './tests/user-test';
-import delay from './utils/delay';
+
+import { Test } from './test';
+import Factories from './tests';
 import destroyActors from './utils/destroyActors';
 
+const SuccessColor = MRE.Color3.Green();
+const FailureColor = MRE.Color3.Red();
+const NeutralColor = MRE.Color3.Yellow();
+
 /**
- * Functional Test Application.
+ * Functional Test Application. Takes query arguments to the websocket connection
+ * @param test - Initialize menu on a particular test
+ * @param autostart - Start the test immediately on user join
+ * @param nomenu - Do not spawn the controls
  */
 export default class App {
-    private activeTests: { [id: string]: Test } = {};
     private _rpc: MRERPC.ContextRPC;
-    private firstUser: MRESDK.User;
-    private _connectedUsers: { [id: string]: MRESDK.User } = {};
+    private firstUser: MRE.User;
+    private _connectedUsers: { [id: string]: MRE.User } = {};
+    private testNames = Object.keys(Factories).sort();
+    private testResults: { [name: string]: boolean } = {};
+
+    private activeTest: Test = null;

```

Comment:
>activeTest [](start = 12, length = 10)

As we discussed offline: When the functional-test app is restarted and reloads into the scene, we need to be able to immediately click on the same point in space to relaunch the same test as before the restart. This is important for rapid dev/test iteration when developing a test. Is that still supported?

Reply:
The query argument `test` will skip to that item in the menu, so yes. There are also `autorun` and `nomenu` arguments available.

Reply:
That isn't the same. I don't want the test to auto start.

---
In reply to: [258618909](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#discussion_r258618909) [](ancestors = 258618909)

Reply:
In a single session, we must be able to:
- click on the same spot in space to start the test after restarting the app.
- explicitly start the test (not auto start).
- run the test multiple times.

---
In reply to: [258619934](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#discussion_r258619934) [](ancestors = 258619934,258618909)

Reply:
Maybe I need button labels. Set your URI to `ws://localhost:3901/?test=my-test`, and the app will start on your test, ready to go when you click Start (the bottom button).

---
In reply to: [258620721](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#discussion_r258620721) [](ancestors = 258620721,258619934,258618909)